### PR TITLE
Feature: adds tags to DynamoDB configs to tag tables

### PR DESCRIPTION
**Pull request by epps**
# Description

This PR introduces the ability to tag DynamoDB tables created by Featureform to track usage and spending. See [here](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tagging.html) for tagging details.

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (non-breaking change that modifies some code)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

## Summary by CodeRabbit

- **New Features**
	- DynamoDB resource registration now accepts custom table tags, allowing for additional metadata during setup.
	- Enhanced tag validation ensures tags adhere to proper naming and formatting rules.
	- Default ownership tagging has been added to streamline resource organization.
	- New field for tags added to DynamoDB configuration for improved metadata management.
- **Bug Fixes**
	- Improved error handling for invalid tag structures during DynamoDB configuration.
- **Tests**
	- Added comprehensive tests for the new tagging functionality in DynamoDB, covering various scenarios and edge cases.

### CODE REVIEW

File: client/src/featureform/resources.py [hunk 1]
```diff
@@ -741,6 +750,49 @@ def __eq__(self, __value: object) -> bool:
 
         return self.region == __value.region
 
+    def validate_table_tags(self, table_tags: Optional[dict]) -> list:
```
Comment (aolfat) [inline client/src/featureform/resources.py:753]: does the equals need to be updated? If not please delete the equals. The fact that we added back equals methods on dataclasses is really bad

File: client/src/featureform/register.py [hunk 2]
```diff
@@ -3300,8 +3300,9 @@ def register_dynamodb(
         region: str,
         description: str = "",
         team: str = "",
-        tags: List[str] = [],
-        properties: dict = {},
+        tags: Optional[List[str]] = None,
+        properties: Optional[dict] = None,
+        table_tags: Optional[dict] = None,
```
Comment (aolfat) [inline client/src/featureform/register.py:3305]: nice

**Review by aolfat — APPROVED** [review]

File: client/src/featureform/resources.py (hunk 1 above)
Comment (epps) [reply client/src/featureform/resources.py:753]: Removing equals method instead after discussing dataclass already covering comparison.

File: client/src/featureform/client.py [hunk 3]
```diff
@@ -832,26 +832,7 @@ def get_dynamodb(self, name):
...
-        ):
-            credentials = AWSAssumeRoleCredentials()
-        else:
-            raise ValueError("Invalid Credentials Type")
-
-        dynamodb_config = DynamodbConfig(
-            region=deserialized_config["Region"],
-            credentials=credentials,
-        )
+        dynamodb_config = DynamodbConfig.deserialize(provider.serialized_config)
```
Comment (aolfat) [inline client/src/featureform/client.py:835]: yay
